coprocessor/fhevm-engine/fhevm-listener/src/health_check.rs

```diff
@@ -0,0 +1,238 @@
+use std::net::SocketAddr;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use std::sync::Arc;
+
+use alloy::providers::{ProviderBuilder, WsConnect};
+
+use axum::{
+    extract::State,
+    http::StatusCode,
+    response::{IntoResponse, Json},
+    routing::get,
+    Router,
+};
+use serde::Serialize;
+use sqlx::postgres::{PgConnectOptions, PgPoolOptions};
+use tracing::{error, info};
+use tokio::net::TcpListener;
+use tokio::sync::RwLock;
+use tokio_util::sync::CancellationToken;
+
+
+
+/// Represents the health status of the transaction sender service
+#[derive(Clone, Debug, Serialize)]
+pub struct Health {
+    pub status_code: u16,
+    pub status: &'static str,
+    /// Any error is unhealthy
+    pub healthy: bool,
+    /// Database connection status
+    pub database_connected: bool,
+    /// Blockchain provider connection status
+    pub blockchain_connected: bool,
+    pub message: String,
+    pub last_tick: u64,
+}
+
+const HEALTHY: &str = "healthy";
+const UNHEALTHY: &str = "unhealthy";
+const OK: u16 = StatusCode::OK.as_u16();
+const INTERNAL_SERVER_ERROR: u16 = StatusCode::INTERNAL_SERVER_ERROR.as_u16();
+
+impl Health {
+
+    pub fn initial() -> Self {
+        Self {
+            database_connected: false,
+            blockchain_connected: false,
+            message: "Not connected".to_string(),
+            status_code: OK,
+            status: UNHEALTHY,
+            healthy: false,
+            last_tick: 0,
+        }
+    }
+
+    pub fn connected(&mut self) {
+        self.status = HEALTHY;
+        self.message = "Only connected".to_string();
+        self.status_code = OK;
+        self.healthy = true;
+        self.database_connected = true;
+        self.blockchain_connected = true;
+        self.tick();
+    }
+
+    pub fn reset(&mut self) {
+        if self.last_tick == 0 {
+            return;
+        }
+        self.status = HEALTHY;
+        self.status_code = OK;
+        self.healthy = true;
+        self.database_connected = true;
+        self.blockchain_connected = true;
+        self.message = "".to_string();
+    }
+
+    pub fn tick(&mut self) {
+        self.last_tick = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
+    }
+
+    pub fn unhealthy(&mut self) {
+        self.healthy = false;
+        self.status = UNHEALTHY;
+        self.status_code = INTERNAL_SERVER_ERROR;
+    }
+
+    pub fn check_last_tick(&mut self) {
+        if self.last_tick == 0 {
+            // not connected yet
+            return;
+        }
+        let timestamp_now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
+        let elapsed_time = timestamp_now - self.last_tick;
+        if elapsed_time > 30 {
+            self.message += &format!("Last tick is too old {elapsed_time}.\n");
+            self.unhealthy();
+        }
+    }
+
+    pub async fn check_database_connected(&mut self, database_url: &str) {
+        let Ok(options) = database_url.parse::<PgConnectOptions>() else {
```

Comment:
Can you reuse the DB pool that is in use?

Reply:
Same answer as provider.